File: src/mystwin.cc
#include "mystwin.h"

struct WindowRectCache {
	RECT rect;
	HWND parent;
};
std::unordered_map<HWND, WindowRectCache> cacheMap;

bool mystwin::Init(const Napi::Env env) {
    auto progman = FindWindow(L"Progman", NULL);
    auto result = SendMessageTimeout(progman, WM_SPAWN_WORKER, 0, 0, SMTO_NORMAL, 1000, NULL);

	if (!result) {
		Napi::TypeError::New(env, "Cannot send message timeout.").ThrowAsJavaScriptException();
		return false;
	}

	EnumWindows([](HWND hwnd, LPARAM lParam) -> BOOL {
        shelldll = FindWindowEx(hwnd, NULL, L"SHELLDLL_DefView", NULL);
        if (shelldll) {
            *reinterpret_cast<HWND*>(lParam) = FindWindowEx(NULL, hwnd, L"WorkerW", NULL);
            return FALSE;
        }
        return TRUE;
    }, reinterpret_cast<LPARAM>(&workerw));

	if (shelldll == NULL) {
		shelldll = FindWindowEx(progman, NULL, L"SHELLDLL_DefView", NULL);
	}
	if (workerw == NULL) {
		workerw = FindWindowEx(progman, NULL, L"WorkerW", NULL);
	}

	return shelldll != NULL && workerw != NULL;
}

Napi::Value mystwin::Refresh(const Napi::CallbackInfo& info, std::string action) {
	auto env = info.Env();

	if (!info[0].IsBuffer()) {
		Napi::TypeError::New(env, "Invalid arguments").ThrowAsJavaScriptException();
		return env.Null();
	}

	auto windowHandleBuffer = info[0].As<Napi::Buffer<uint8_t>>().Data();
	LONG_PTR handle = *reinterpret_cast<LONG_PTR*>(windowHandleBuffer);
	HWND hwnd = (HWND)(LONG_PTR)handle;

	if (!Init(env)) {
		Napi::TypeError::New(env, "System Windows not found.").ThrowAsJavaScriptException();
		return env.Null();
	}

	auto& cache = cacheMap[hwnd];

	RECT rect;
	GetWindowRect(hwnd, &rect);

	HWND insert = NULL;

	UINT flags = SWP_NOZORDER;

	int xPos = rect.left;
	int yPos = rect.top;
	int width = rect.right - rect.left;
	int height = rect.bottom - rect.top;

	uint32_t opacityValue = 255;

	LONG style = GetWindowLong(hwnd, GWL_STYLE);
	LONG exStyle = GetWindowLong(hwnd, GWL_EXSTYLE);

	style |= WS_TILEDWINDOW;
	exStyle |= WS_EX_LAYERED;

	if (info.Length() == 1) {
		if (action == "Close") {
			DestroyWindow(hwnd);\
			return env.Null();
		} else if (action == "Minimize") {
			ShowWindow(hwnd, SW_MINIMIZE);
			return env.Null();
		} else if (action == "Maximize") {
			ShowWindow(hwnd, SW_MAXIMIZE);
			return env.Null();
		} else if (action == "AttachAsWallpaper") {
			cache.parent = workerw;
		} else if (action == "AttachAsDesktop") {
			cache.parent = shelldll;
		} else if (action == "AttachToTop") {
			insert = HWND_TOPMOST;
			flags = NULL;
		} else if (action == "DetachWindow") {
			insert = HWND_NOTOPMOST;
			flags = NULL;
			cache.parent = NULL;
		} else if (action == "SendToBack") {
			insert = HWND_BOTTOM;
			flags = NULL;
		} else if (action == "BringToFront") {
			insert = HWND_TOP;
			flags = NULL;
		}
	} else if (info.Length() == 2) {
		if (action == "SetOpacity") {
			opacityValue = std::clamp<uint32_t>(info[1].As<Napi::Number>().Uint32Value(), 0u, 255u);
		} else {
			bool enable = info[1].As<Napi::Boolean>().Value();
			if (action == "ToggleFrame") {
				if (enable) {
					style |= (WS_CAPTION | WS_THICKFRAME | WS_SYSMENU);
				} else {
					style &= ~(WS_CAPTION | WS_THICKFRAME | WS_SYSMENU);
				}
			} else if (action == "ToggleTaskBar") {
				if (enable) {
					exStyle |= WS_EX_TOOLWINDOW;
				} else {
					exStyle &= ~WS_EX_TOOLWINDOW;
				}
			} else if (action == "ToggleOverlay") {
				if (enable) {
					exStyle |= WS_EX_TRANSPARENT;
				} else {
					exStyle &= ~WS_EX_TRANSPARENT;
				}
			} else if (action == "ToggleFullScreen") {
				if (enable) {
					GetWindowRect(hwnd, &cache.rect);
					MONITORINFO mi = { sizeof(mi) };
					if (GetMonitorInfo(MonitorFromWindow(hwnd, MONITOR_DEFAULTTOPRIMARY), &mi)) {
						xPos = mi.rcMonitor.left;
						yPos = mi.rcMonitor.top;
						width = mi.rcMonitor.right - mi.rcMonitor.left;
						height = mi.rcMonitor.bottom - mi.rcMonitor.top;
					}
				} else {
					xPos = cache.rect.left;
					yPos = cache.rect.top;
					width = cache.rect.right - cache.rect.left;
					height = cache.rect.bottom - cache.rect.top;
				}
			}
		}
	} else if (info.Length() == 3) {
		if (action == "MoveWindow") {
			xPos = info[1].As<Napi::Number>().Int32Value();
			yPos = info[2].As<Napi::Number>().Int32Value();
		} else if (action == "ResizeWindow") {
			width = info[1].As<Napi::Number>().Int32Value();
			height = info[2].As<Napi::Number>().Int32Value();
		}
	} else {
		Napi::TypeError::New(env, "Invalid arguments").ThrowAsJavaScriptException();
		return env.Null();
	}

	BYTE opacity = static_cast<BYTE>(opacityValue);

	SetParent(hwnd, cache.parent);
	SetWindowLong(hwnd, GWL_STYLE, style);
	SetWindowLong(hwnd, GWL_EXSTYLE, exStyle);
	SetWindowPos(hwnd, NULL, 0, 0, 0, 0, SWP_NOZORDER | SWP_FRAMECHANGED);
	SetWindowPos(hwnd, insert, xPos, yPos, width, height, flags);
	SetLayeredWindowAttributes(hwnd, 0, opacity, LWA_ALPHA);

	return env.Null();
}

Napi::Value mystwin::AttachAsWallpaperExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "AttachAsWallpaper");
}

Napi::Value mystwin::AttachAsDesktopExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "AttachAsDesktop");
}

Napi::Value mystwin::AttachToTopExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "AttachToTop");
}

Napi::Value mystwin::DetachWindowExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "DetachWindow");
}

Napi::Value mystwin::SendToBackExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "SendToBack");
}

Napi::Value mystwin::BringToFrontExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "BringToFront");
}

Napi::Value mystwin::MoveWindowExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "MoveWindow");
}

Napi::Value mystwin::ResizeWindowExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "ResizeWindow");
}

Napi::Value mystwin::SetOpacityExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "SetOpacity");
}

Napi::Value mystwin::ToggleFrameExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "ToggleFrame");
}

Napi::Value mystwin::ToggleTaskBarExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "ToggleTaskBar");
}

Napi::Value mystwin::ToggleOverlayExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "ToggleOverlay");
}

Napi::Value mystwin::ToggleFullScreenExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "ToggleFullScreen");
}

Napi::Value mystwin::MaximizeExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "Maximize");
}

Napi::Value mystwin::MinimizeExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "Minimize");
}

Napi::Value mystwin::CloseExport(const Napi::CallbackInfo& info) {
	return Refresh(info, "Close");
}

File: lib/mystwin.js
// @ts-check

/**
 * @module mystwin
 */

'use strict'

const bindings = require('bindings')
const mystwin = bindings('mystwin')

/**
 * @summary Attaches a window into the Desktop Wallpaper layer (1 layer above wallpaper).
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.attachAsWallpaper(currentWindow);
 */
exports.attachAsWallpaper = function attachAsWallpaper (window) {
  mystwin.attachAsWallpaper(window.getNativeWindowHandle())
}

/**
 * @summary Attaches a window into the Windows Desktop layer (1 layer above icons).
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.attachAsDesktop(currentWindow);
 */
exports.attachAsDesktop = function attachAsDesktop (window) {
  mystwin.attachAsDesktop(window.getNativeWindowHandle())
}

/**
 * @summary Attaches a window into the Top Most layer (Always On Top).
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.attachToTop(currentWindow);
 */
exports.attachToTop = function attachToTop (window) {
  mystwin.attachToTop(window.getNativeWindowHandle())
}

/**
 * @summary Removes a child window from its parent, making it independent again.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.detachWindow(currentWindow);
 */
exports.detachWindow = function detachWindow (window) {
  mystwin.detachWindow(window.getNativeWindowHandle())
}

/**
 * @summary Sends a window to the bottom of the Z-order.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.sendToBack(currentWindow);
 */
exports.sendToBack = function sendToBack (window) {
  mystwin.sendToBack(window.getNativeWindowHandle())
}

/**
 * @summary Brings a window to the top of the Z-order.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.bringToFront(currentWindow);
 */
exports.bringToFront = function bringToFront (window) {
  mystwin.bringToFront(window.getNativeWindowHandle())
}

/**
 * @summary Moves a window to the specific position without resizing.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 * @param {number} x
 * @param {number} y
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.moveWindow(currentWindow);
 */
exports.moveWindow = function moveWindow (window, x, y) {
  mystwin.moveWindow(window.getNativeWindowHandle(), x, y)
}

/**
 * @summary Resizes a window to the given dimensions without moving.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 * @param {number} width
 * @param {number} height
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.resizeWindow(currentWindow);
 */
exports.resizeWindow = function resizeWindow (window, width, height) {
  mystwin.resizeWindow(window.getNativeWindowHandle(), width, height)
}

/**
 * @summary Sets the window opacity (0 = fully transparent, 255 = fully opaque).
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 * @param {number} opacity
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.setOpacity(currentWindow);
 */
exports.setOpacity = function setOpacity (window, opacity) {
  mystwin.setOpacity(window.getNativeWindowHandle(), opacity)
}

/**
 * @summary Show icon on task bar or not.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 * @param {boolean} enable
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.toggleTaskBar(currentWindow);
 */
exports.toggleTaskBar = function toggleTaskBar (window, enable) {
  mystwin.toggleTaskBar(window.getNativeWindowHandle(), enable)
}

/**
 * @summary Display window's frame or not.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 * @param {boolean} enable
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.toggleFrame(currentWindow);
 */
exports.toggleFrame = function toggleFrame (window, enable) {
  mystwin.toggleFrame(window.getNativeWindowHandle(), enable)
}

/**
 * @summary Allows clicks to pass through a window or not.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 * @param {boolean} enable
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.toggleOverlay(currentWindow);
 */
exports.toggleOverlay = function toggleOverlay (window, enable) {
  mystwin.toggleOverlay(window.getNativeWindowHandle(), enable)
}

/**
 * @summary Display window's in fullscreen mode or not.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 * @param {boolean} enable
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.toggleFullScreen(currentWindow);
 */
exports.toggleFullScreen = function toggleFullScreen (window, enable) {
  mystwin.toggleFullScreen(window.getNativeWindowHandle(), enable)
}

/**
 * @summary Forcely Maximize Window.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.maximize(currentWindow);
 */
exports.maximize = function maximize (window) {
  mystwin.maximize(window.getNativeWindowHandle())
}

/**
 * @summary Forcely Minimize Window.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.minimize(currentWindow);
 */
exports.minimize = function minimize (window) {
  mystwin.minimize(window.getNativeWindowHandle())
}

/**
 * @summary Forcely Close Window.
 * @function
 * @public
 *
 * @param {import('electron').BrowserWindow} window
 *
 * @example
 * const MystWin = require('mystwin');
 * const currentWindow = require('electron').remote.getCurrentWindow();
 *
 * MystWin.close(currentWindow);
 */
exports.close = function close (window) {
  mystwin.close(window.getNativeWindowHandle())
}
